### ISSUE TITLE
Default nuxt app does not work out of the box.

### ISSUE THREAD

**Issue by jose-fully-ported** [heroku/buildpacks-nodejs#712]
There are a few issues here (maybe all related to pnpm somehow) but:

```shell
# create a new nuxt app
npx nuxi@latest init test-app

# setup a pnpm lock file
pnpm install

# run the build
# will fail because the `packageManager` isn't set
pack build --builder heroku/builder:22 test-app

# set the packageManager
contents="$(jq '.packageManager = "pnpm@8.10.2"' package.json)"
echo -E "${contents}" > package.json

# run the build
# will fail because it can't find a dependency
pack build --builder heroku/builder:22 test-app
```

Build error:
```
 ERROR  Cannot find package '@vue/compiler-sfc' imported from /layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.7/node_modules/unplugin-vue-router/dist/chunk-QVOU6AJO.mjs

  at new NodeError (node:internal/errors:406:5)
  at packageResolve (node:internal/modules/esm/resolve:789:9)
  at moduleResolve (node:internal/modules/esm/resolve:838:20)
  at defaultResolve (node:internal/modules/esm/resolve:1043:11)
  at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:383:12)
  at ModuleLoader.resolve (node:internal/modules/esm/loader:352:25)
  at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:228:38)
  at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39)
  at link (node:internal/modules/esm/module_job:84:36)



 ERROR  Cannot find package '@vue/compiler-sfc' imported from /layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.7/node_modules/unplugin-vue-router/dist/chunk-QVOU6AJO.mjs
```

Aside, `pnpm install` works fine out of the buildpack.

Note that packeto and the old heroku buildpack do not support pnpm, so there isn't a way to switch to another builder.

**Comment by colincasey** [comment]
@jose-fully-ported thanks for submitting this issue!  

I was able to reproduce your issue with the default Nuxt app + pnpm. This will need to be investigated.

> I think one thing the buildpack should do is pass buildpack detection when no packageManager is set but a lockfile is detected but add a warning if so. Ditto if no lockfile but there is a package.json, just show a warning.

We're planning to relax the detection routines for several of the Node.js buildpacks so we can provide warning or errors later in the process.

**Comment by colincasey** [comment]
@jose-fully-ported Quick update here.  I modified the `postinstall` script to run `NODE_DEBUG=module nuxt prepare` and got the following output (formatted for readability):

```
MODULE 311: looking for "@vue/compiler-sfc" in [
  "/layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.8/node_modules/unplugin-vue-router/dist/chunk-QVOU6AJO.mjs/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.8/node_modules/unplugin-vue-router/dist/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.8/node_modules/unplugin-vue-router/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/unplugin-vue-router@0.7.0_vue-router@4.2.5_vue@3.3.8/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/node_modules",
  "/layers/heroku_nodejs-pnpm-install/node_modules",
  "/layers/node_modules",
  "/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/nuxt@3.8.1_vite@4.5.0/node_modules/nuxt/bin/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/nuxt@3.8.1_vite@4.5.0/node_modules/nuxt/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/nuxt@3.8.1_vite@4.5.0/node_modules",
  "/layers/heroku_nodejs-pnpm-install/virtual/node_modules",
  "/home/heroku/.node_modules",
  "/home/heroku/.node_libraries",
  "/layers/heroku_nodejs-engine/dist/lib/node"
]
```

You'll notice that the `/workspace/node_modules` directory isn't in that list which is why the module load fails even though it is installed. 

I've yet to narrow down why the `/workspace/node_modules` directory is not in the module load path and if that's caused by our pnpm setup, a bug in pnpm or nuxt, or something else.

**Comment by colincasey** [comment]
* Using `yarn@3.6.4`, no (tried both `nodeLinker: "pnp"` and `nodeLinker: "node_modules"`)
* Using `npm@10.1.0`, yes.

I'm also able to reproduce the `pnpm` error outside of the buildpack by copying how the buildpack configures `pnpm`:
* `pnpm config set store-dir /tmp/pnpm/store`
* `pnpm config set virtual-store-dir /tmp/pnpm/virtual`
* `pnpm install`

The above steps give the same `ERROR  Cannot find package '@vue/compiler-sfc'` seen in the buildpack.

**Comment by colincasey** [comment]
@joshwlewis The pnpm docs around the [`virtual-store-dir` configuration](https://pnpm.io/npmrc#virtual-store-dir) appear to be misleading. It gives an example that suggests the `virtual-store-dir` could be located outside of the project directory but given the comment [here](https://github.com/pnpm/pnpm/issues/4819), this may not actually work:

> Packages only search for modules in node_modules folders that are in parent directories. If you move out the virtual store from your project's directory, where the root modules are symlinked, the packages will not look for it.

I experimented with leaving the `virtual-store-dir` as the default (`node_modules/.pnpm`) and this fixes the issue locally but, in the Docker environment, it ends up producing a lot of "cross-device link not permitted" errors.

**Comment by joshwlewis** [comment]
> Would it be possible to do something like... Create the node_modules dir in a not-volume path and then move it into place for use?

Yeah, this is pretty similar to the pattern we have in the classic buildpack caching. This would certainly have a performance cost, since we're copying data across mounts. 

I've gotten something working locally, where I symlink `{virtual_store_layer}/node_modules` to `/workspace/node_modules`, so that the module resolver will be able to find `/workspace/node_modules` from `{virtual_store_layer}/cache/some-dependency` by traversing up parent directories.

**Comment by colincasey** [comment]
Closing.  Should be fixed by #737
